src/discordcr/voice.cr

```diff
@@ -0,0 +1,268 @@
+
+    def connect(endpoint : String, port : UInt32, ssrc : UInt32)
+      endpoint = endpoint[6..-1] if endpoint.starts_with? "wss://"
+      endpoint = endpoint.gsub(":80", "") # The endpoint may contain a port, we don't want that
+
+      @ssrc = ssrc
+      @socket.connect(endpoint, port)
+    end
+
+    # Sends a discovery packet to Discord, telling them that we want to know our
+    # IP so we can select the protocol on the VWS
+    def send_discovery
+      io = IO::Memory.new(70)
+
+      io.write_bytes(@ssrc.not_nil!, IO::ByteFormat::BigEndian) # [@ssrc].pack("N")
+      io.write(Bytes.new(66, 0_u8))                             # 66 null bytes for a total of 70 bytes
+
+      @socket.write(io.to_slice)
+    end
+
+    # Awaits a response to the discovery request and returns our local IP and
+    # port once the response is received
+    def receive_discovery_reply : {String, UInt16}
+      buf = Bytes.new(70)
+      @socket.receive(buf)
+      io = IO::Memory.new(buf)
+
+      io.seek(4)                                              # The first four bytes are just the SSRC again, we don't care about that
+      ip = io.read_string(64).delete("\0")                    # Our IP
+      port = io.read_bytes(UInt16, IO::ByteFormat::BigEndian) # Our port
+
+      {ip, port}
+    end
+
+    # Sends 20 ms of opus audio data to Discord, with the specified sequence and
+    # time (used on the receiving client to synchronise packets)
+    def send_audio(buf, sequence, time)
+      header = create_header(sequence, time)
+
+      buf = encrypt_audio(header, buf)
+
+      new_buf = Bytes.new(header.size + buf.size)
+      header.copy_to(new_buf)
+      buf.copy_to(new_buf + header.size)
+
+      @socket.write(new_buf)
+    end
+
+    private def create_header(sequence : UInt16, time : UInt32) : Bytes
+      io = IO::Memory.new(12)
+
+      # Write the magic bytes required by Discord
+      io.write_byte(0x80_u8)
+      io.write_byte(0x78_u8)
+
+      # Write the actual information in the header
+      io.write_bytes(sequence, IO::ByteFormat::BigEndian)
+      io.write_bytes(time, IO::ByteFormat::BigEndian)
+      io.write_bytes(@ssrc.not_nil!, IO::ByteFormat::BigEndian)
+
+      io.to_slice
+    end
+
+    private def encrypt_audio(header : Bytes, buf : Bytes) : Bytes
+      raise "No secret key was set!" unless @secret_key
+
+      nonce = Bytes.new(24, 0_u8) # 24 null bytes
+      header.copy_to(nonce)       # First 12 bytes of nonce is the header
+
+      # Sodium constants
+      zero_bytes = Sodium.crypto_secretbox_xsalsa20poly1305_zerobytes
+      box_zero_bytes = Sodium.crypto_secretbox_xsalsa20poly1305_boxzerobytes
+
+      # Prepend the buf with zero_bytes zero bytes
+      message = Bytes.new(buf.size + zero_bytes, 0_u8)
+      buf.copy_to(message + zero_bytes)
+
+      # Create a buffer for the ciphertext
+      c = Bytes.new(message.size)
+
+      # Encrypt
+      key = Bytes.new(@secret_key.not_nil!.to_unsafe, @secret_key.not_nil!.size)
```

Comment:
Well, they don't. Discord sends it as an array so it will fundamentally be an array somewhere, but you're right that it's more efficient to do the to_unsafe thing once rather than for every packet